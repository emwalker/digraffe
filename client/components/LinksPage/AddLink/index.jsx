// @flow
import React, { Component } from 'react'
import { Button, FormGroup, Input } from 'reactstrap'

import createLinkMutation from '../../../mutations/createLinkMutation'

type Props = {
  organization: {
    id: string,
    resourceId: string,
  },
  relay: {
    environment: Object,
  },
}

type State = {
  url: string,
}

class AddLink extends Component<Props, State> {
  state = {
    url: '',
  }

  onSubmit = () => {
    const {
      id: orgId,
      resourceId: organizationResourceId,
    } = this.props.organization
    createLinkMutation(
      this.props.relay.environment,
      orgId,
      {
        organizationResourceId,
        url: this.state.url,
      },
    )
    this.setState({ url: '' })
  }

  updateUrl = (event: Object) => {
    this.setState({ url: event.currentTarget.value })
  }

  render = () => (
    <div className="form-container">
      <FormGroup>
        <Input
          className="link-url test-link-url"
          placeholder="Link URL"
          value={this.state.url}
          onChange={this.updateUrl}
        />
        <Button size="sm" onClick={this.onSubmit}>Add</Button>
      </FormGroup>
    </div>
  )
}

export default AddLink
