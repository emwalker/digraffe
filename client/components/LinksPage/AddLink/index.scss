.form-container {
  text-align: right;
}

.link-url {
  display: inline-block;
  width: 75%;
  margin-right: 1em;
}
