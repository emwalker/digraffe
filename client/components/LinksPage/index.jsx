// @flow
import React from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

// import AddTopic from './AddTopic'
import ListView from '../ui/ListView'
import { liftNodes } from '../../utils'

type Props = {
  organization: {
    links: Object,
  },
}

const LinksPage = ({ organization }: Props) => (
  <ListView
    title="Links"
    items={liftNodes(organization.links)}
  >
    <div>Add link</div>
  </ListView>
)

export const query = graphql`
query LinksPage_query_Query($orgResourceId: String!) {
  viewer {
    ...LinksPage_viewer
  }

  organization(resourceId: $orgResourceId) {
    ...LinksPage_organization
  }
}`

export default createFragmentContainer(LinksPage, graphql`
  fragment LinksPage_viewer on User {
    name
  }

  fragment LinksPage_organization on Organization {
    id
    resourceId

    links(first: 100) @connection(key: "Organization_links") {
      edges {
        node {
          id
          display: title
          resourcePath
        }
      }
    }
  }
`)
