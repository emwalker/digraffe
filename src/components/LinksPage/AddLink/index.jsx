// @flow
import React, { Component } from 'react'
import { FormGroup, Input } from 'reactstrap'
import { graphql, createFragmentContainer } from 'react-relay'
import Select from 'react-select'

import createLinkMutation from '../../../mutations/createLinkMutation'
import { liftNodes } from '../../../utils'

type Props = {
  organization: {
    id: string,
    resourceId: string,
    availableTopics: Array<{
      label: string,
      value: string,
    }>,
  },
  relay: {
    environment: Object,
  },
}

type State = {
  selectedOption: string,
  url: string,
}

class AddLink extends Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.options = liftNodes(props.organization.availableTopics)
  }

  state = {
    selectedOption: '',
    url: '',
  }

  onInputChange = (event: Object) => {
    this.setState({ url: event.currentTarget.value })
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.setState(({ url }) => {
        this.addUrl(url)
        return { url: '' }
      })
  }

  onSelect = (selectedOption) => {
    this.setState({ selectedOption })
  }

  addUrl = (url) => {
    const {
      id: orgId,
      resourceId: organizationResourceId,
    } = this.props.organization
    createLinkMutation(
      this.props.relay.environment,
      orgId,
      { organizationResourceId, url },
    )
  }

  options: Object[]

  render = () => (
    <div className="form-container">
      <FormGroup>
        <Select
          id="link-topic-id"
          name="link-topic"
          onChange={this.onSelect}
          options={this.options}
          placeholder="Select a topic"
          value={this.state.selectedOption}
        />
      </FormGroup>
      <FormGroup>
        <Input
          className="link-url test-link-url"
          onChange={this.onInputChange}
          onKeyPress={this.onKeyPress}
          placeholder="Link URL"
          value={this.state.url}
        />
      </FormGroup>
    </div>
  )
}

export default createFragmentContainer(AddLink, graphql`
  fragment AddLink_organization on Organization {
    id
    resourceId

    availableTopics: topics(first: 100) {
      edges {
        node {
          label: name
          value: resourcePath
        }
      }
    }
  }
`)
