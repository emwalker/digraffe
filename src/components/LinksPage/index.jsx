// @flow
import React from 'react'
import { graphql, createFragmentContainer } from 'react-relay'
import { isEmpty } from 'ramda'

import type { OrganizationType } from '../types'
import { liftNodes } from '../../utils'
import Link from '../ui/Link'
import List from '../ui/List'

type Props = {
  organization: OrganizationType,
}

const LinksPage = ({ organization, ...props }: Props) => {
  const links = liftNodes(organization.links)
  return (
    <div>
      <List
        placeholder="There are no links"
        hasItems={!isEmpty(links)}
      >
        { links.map(link => (
          <Link
            key={link.id}
            organization={organization}
            link={link}
            {...props}
          />
        )) }
      </List>
    </div>
  )
}

export const query = graphql`
  query LinksPage_query_Query($orgId: ID!) {
    organization(id: $orgId) {
      ...LinksPage_organization
    }
  }
`

export default createFragmentContainer(LinksPage, graphql`
  fragment LinksPage_organization on Organization {
    ...Link_organization

    links(first: 1000) @connection(key: "Organization_links") {
      edges {
        node {
          id
          ...Link_link
        }
      }
    }
  }
`)
