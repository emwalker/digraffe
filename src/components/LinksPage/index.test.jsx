import React from 'react'
import { shallow } from 'enzyme'
import LinksPage from './index'

jest.mock('react-relay', () => ({ createFragmentContainer: component => component }))

describe('<LinksPage />', () => {
  const viewer = {
    name: 'Rezrov',
  }

  const view = {
    resourceId: 'organization:tyrell',
    topics: {
      edges: [
        {
          node: {
            name: 'Frotz',
            resourceId: '/topics/1234',
            id: '1234',
          },
        },
      ],
    },
  }

  const relay = {
    environment: {},
  }

  const wrapper = shallow(
    <LinksPage
      view={view}
      viewer={viewer}
      relay={relay}
    />,
  )

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
