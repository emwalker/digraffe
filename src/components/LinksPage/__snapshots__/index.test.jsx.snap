// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinksPage /> renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LinksPage
    organization={
      Object {
        "resourceId": "organization:tyrell",
        "topics": Object {
          "edges": Array [
            Object {
              "node": Object {
                "id": "1234",
                "name": "Frotz",
                "resourceId": "/topics/1234",
              },
            },
          ],
        },
      }
    }
    relay={
      Object {
        "environment": Object {},
      }
    }
    viewer={
      Object {
        "name": "Rezrov",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="Subhead"
        >
          <div
            className="Subhead-heading"
          >
            Links
          </div>
        </div>,
        <Unknown
          items={Array []}
          organization={
            Object {
              "resourceId": "organization:tyrell",
              "topics": Object {
                "edges": Array [
                  Object {
                    "node": Object {
                      "id": "1234",
                      "name": "Frotz",
                      "resourceId": "/topics/1234",
                    },
                  },
                ],
              },
            }
          }
          relay={
            Object {
              "environment": Object {},
            }
          }
          title="Links"
          viewer={
            Object {
              "name": "Rezrov",
            }
          }
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="Subhead-heading"
          >
            Links
          </div>,
          "className": "Subhead",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Links",
            "className": "Subhead-heading",
          },
          "ref": null,
          "rendered": "Links",
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "items": Array [],
          "organization": Object {
            "resourceId": "organization:tyrell",
            "topics": Object {
              "edges": Array [
                Object {
                  "node": Object {
                    "id": "1234",
                    "name": "Frotz",
                    "resourceId": "/topics/1234",
                  },
                },
              ],
            },
          },
          "relay": Object {
            "environment": Object {},
          },
          "title": "Links",
          "viewer": Object {
            "name": "Rezrov",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="Subhead"
          >
            <div
              className="Subhead-heading"
            >
              Links
            </div>
          </div>,
          <Unknown
            items={Array []}
            organization={
              Object {
                "resourceId": "organization:tyrell",
                "topics": Object {
                  "edges": Array [
                    Object {
                      "node": Object {
                        "id": "1234",
                        "name": "Frotz",
                        "resourceId": "/topics/1234",
                      },
                    },
                  ],
                },
              }
            }
            relay={
              Object {
                "environment": Object {},
              }
            }
            title="Links"
            viewer={
              Object {
                "name": "Rezrov",
              }
            }
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="Subhead-heading"
            >
              Links
            </div>,
            "className": "Subhead",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Links",
              "className": "Subhead-heading",
            },
            "ref": null,
            "rendered": "Links",
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "items": Array [],
            "organization": Object {
              "resourceId": "organization:tyrell",
              "topics": Object {
                "edges": Array [
                  Object {
                    "node": Object {
                      "id": "1234",
                      "name": "Frotz",
                      "resourceId": "/topics/1234",
                    },
                  },
                ],
              },
            },
            "relay": Object {
              "environment": Object {},
            },
            "title": "Links",
            "viewer": Object {
              "name": "Rezrov",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
